File: PythonSolutions/Neetcode150/binarytree_problems/1448_count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        
        self.count = 0

        # Preorder traversal
        def dfs(node, maxInPath):
            if not node:
                return maxInPath
            
            # Assign the next greater value along the path of DFS
            if node.val >= maxInPath:
                self.count += 1
                maxInPath = node.val
            
            # Pass the current maxInPath (or new maxInPath) along left and right subtrees
            dfs(node.left, maxInPath)
            dfs(node.right, maxInPath)

            # After left and right subtree traversal propagate the CURRENT maxInPath NOT from subtrees
            return maxInPath
        
        dfs(root, float("-inf"))
        return self.count

File: TypescriptSolutions/Neetcode150/binary-tree_problems/1448_count-good-nodes-in-binary-tree.ts
export {};

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}


const goodNodes = (root: TreeNode | null): number => {
    if (!root) return 0;

    let count = 0;

    const dfs = (node, maxInPath): number => {
        if (!node) return maxInPath;

        if (node.val >= maxInPath) {
            count++;
            maxInPath = node.val;
        }

        dfs(node.left, maxInPath);
        dfs(node.right, maxInPath);

        return maxInPath;
    };

    dfs(root, -Infinity);
    return count;
};